aptos-move/block-executor/src/executor.rs

```diff
@@ -311,45 +360,49 @@ where
             Vec<FeeStatement>,
             Option<Error<E::Error>>,
         )>,
+        base_view: &S,
+        shared_counter: &AtomicU32,
+        executor: &E,
+        block: &[T],
     ) {
-        while let Some(txn_idx) = scheduler.try_commit() {
-            let read_set = last_input_output
-                .read_set(txn_idx)
-                .expect("Read set must be recorded");
-            let mut execution_still_valid =
-                read_set.validate_delayed_field_reads(versioned_cache.delayed_fields(), txn_idx);
-
-            match last_input_output.output_category(txn_idx) {
-                Some(ErrorCategory::SpeculativeExecutionError) => {
-                    assert!(!execution_still_valid);
-                },
-                Some(ErrorCategory::CodeInvariantError) => {
-                    panic!();
-                },
-                _ => (),
-            };
+        while let Some((txn_idx, next_incarnation)) = scheduler.try_commit() {
+            if !Self::validate_commit_ready(txn_idx, versioned_cache, last_input_output) {
+                // Transaction needs to be re-executed, one final time.
 
-            if execution_still_valid {
-                if let Some(delayed_field_ids) = last_input_output.delayed_field_keys(txn_idx) {
-                    if let Err(e) = versioned_cache
-                        .delayed_fields()
-                        .try_commit(txn_idx, delayed_field_ids.collect())
-                    {
-                        match e {
-                            CommitError::ReExecutionNeeded(_) => {
-                                execution_still_valid = false;
-                            },
-                            CommitError::CodeInvariantError(_) => {
-                                // TODO: fallback to sequential execution
-                                panic!();
-                            },
-                        }
-                    }
+                Self::update_transaction_failed_validation(
+                    txn_idx,
+                    last_input_output,
+                    versioned_cache,
+                );
+                scheduler.finish_abort_during_commit(txn_idx);
+
+                let (record_succeeded, updates_outside) = Self::execute(
+                    txn_idx,
+                    next_incarnation,
+                    block,
+                    last_input_output,
+                    versioned_cache,
+                    executor,
+                    base_view,
+                    ParallelState::new(versioned_cache, scheduler, shared_counter),
+                );
+                if record_succeeded {
+                    scheduler.finish_execution_during_commit(txn_idx, updates_outside);
+                } else {
+                    // TODO add support
+                    panic!("module re-validation in commit retry");
```

Comment:
This is actually possible in multiple ways, since re-execution of the transaction can have totally different behavior - it could now use a new module that was previously published in the same block. Even without this, some later txn might have published a new version in the meantime, and re-storing captured reads might actually detect the conflict here for the first time. So we shouldn't panic and keep the same handler, I think.

Reply:
there is TODO above :)

intentionally not doing it now (i.e. this method doesn't return Task), need to sync with you how to do it correctly throughput